plugins/outputs/application_insights/application_insights.go

```diff
@@ -0,0 +1,427 @@
+	telemetry.Value = telemetryValue
+	telemetry.Count = telemetryCount
+	telemetry.Properties = metric.Tags()
+	a.addContextTags(metric, telemetry)
+	telemetry.Timestamp = metric.Time()
+
+	// We attempt to set min, max, variance and stddev fields but do not really care if they are not present--
+	// they are not essential for aggregate metric.
+	// By convention AppInsights prefers stddev over variance, so to be consistent, we test for stddev after testing for variance.
+	setFloat64TelemetryProperty(&(telemetry.Min), []string{"min"}, metric, &usedFields)
+	setFloat64TelemetryProperty(&(telemetry.Max), []string{"max"}, metric, &usedFields)
+	setFloat64TelemetryProperty(&(telemetry.Variance), []string{"variance"}, metric, &usedFields)
+	setFloat64TelemetryProperty(&(telemetry.StdDev), []string{"stddev"}, metric, &usedFields)
+
+	return telemetry, usedFields
+}
+
+func (a *ApplicationInsights) createTelemetryForUnusedFields(metric telegraf.Metric, usedFields []string) []appinsights.Telemetry {
+	fields := metric.Fields()
+	retval := make([]appinsights.Telemetry, 0, len(fields))
+
+	for fieldName := range fields {
+		if contains(usedFields, fieldName) {
+			continue
+		}
+
+		telemetry := a.createSimpleMetricTelemetry(metric, fieldName, true)
+		if telemetry != nil {
+			retval = append(retval, telemetry)
+		}
+	}
+
+	return retval
+}
+
+func (a *ApplicationInsights) addContextTags(metric telegraf.Metric, telemetry appinsights.Telemetry) {
+	for tagSourceName, contextTagName := range a.ContextTagSources {
+		contextTagValue, found := metric.GetTag(tagSourceName)
+		if !found {
+			continue
+		}
+
+		telemetry.ContextTags()[contextTagName] = contextTagValue
+	}
+}
+
+func setFloat64TelemetryProperty(
+	telemetryProperty *float64,
+	candidateFields []string,
+	metric telegraf.Metric,
+	usedFields *[]string) error {
+
+	for _, fieldName := range candidateFields {
+		fieldValue, found := metric.GetField(fieldName)
+		if !found {
+			continue
+		}
+
+		metricValue, err := toFloat64(fieldValue)
+		if err != nil {
+			continue
+		}
+
+		*telemetryProperty = metricValue
+
+		if usedFields != nil {
+			*usedFields = append(*usedFields, fieldName)
+		}
+
+		return nil
+	}
+
+	return fmt.Errorf("No field from the candidate list was found in the metric")
+}
+
+func setIntTelemetryProperty(
+	telemetryProperty *int,
+	candidateFields []string,
+	metric telegraf.Metric,
+	usedFields *[]string) error {
+
+	for _, fieldName := range candidateFields {
+		fieldValue, found := metric.GetField(fieldName)
+		if !found {
+			continue
+		}
+
+		metricValue, err := toInt(fieldValue)
+		if err != nil {
+			continue
+		}
+
+		*telemetryProperty = metricValue
+
+		if usedFields != nil {
+			*usedFields = append(*usedFields, fieldName)
+		}
+
+		return nil
+	}
+
+	return fmt.Errorf("No field from the candidate list was found in the metric")
+}
+
+func contains(set []string, val string) bool {
+	for _, elem := range set {
```

Comment:
It will throw if set is nil correct?

Reply:
Nope. This is one of the nice things about Golang: nil values are generally useful/harmless. In this case, ranging over nil map is a no-op.

Reply:
Just tested if it's typed then it's fine, but throws if it's untyped nil.

Reply:
I see. I do not think this case applies to the plugin code